### ISSUE TITLE
ReQuEST accuracy evaluation on ImageNet: ArmCL 18.03 vs. TensorFlow 1.7

### ISSUE THREAD

**Issue by psyhtest** [ARM-software/ComputeLibrary#431]
While evaluating [MobileNets using ArmCL](https://github.com/dividiti/ck-request-asplos18-mobilenets-armcl-opencl) for the [ReQuEST@ASPLOS'18](http://cknowledge.org/request-cfp-asplos2018.html) competition, we observed noticeable discrepancies in predictions of ArmCL 18.03 vs. TensorFlow 1.7. We used the same [pretrained MobileNets-v1 weights shared by Google in 2017](https://github.com/tensorflow/models/blob/1630da3434974e9ad5a0b6d887ac716a97ce03d3/research/slim/nets/mobilenet_v1.md).

For example, the standard accuracy metrics (Top 1 / Top 5 on the [ImageNet](www.image-net.org) validation set of 50,000 images in percent) of 4 models (with the width multiplier of 1.0) are given in the following tables.

### Top 1
Model | TensorFlow 1.1 (?) ([claimed](https://github.com/tensorflow/models/blob/1630da3434974e9ad5a0b6d887ac716a97ce03d3/research/slim/nets/mobilenet_v1.md#pre-trained-models)) | TensorFlow 1.7 (measured) | ArmCL 18.03 (measured)
-|-|-|-
MobileNet_v1_1.0_224 | 70.700 | 70.466 | 70.464 | 
MobileNet_v1_1.0_192 | 69.300 | 68.824 | 68.830 |
MobileNet_v1_1.0_160 | 67.200 | 66.504 | 66.458 |
MobileNet_v1_1.0_128 | 64.100 | 63.580 | 63.586 |

### Top 5
Model | TensorFlow 1.1 (?) ([claimed](https://github.com/tensorflow/models/blob/1630da3434974e9ad5a0b6d887ac716a97ce03d3/research/slim/nets/mobilenet_v1.md#pre-trained-models)) | TensorFlow 1.7 (measured) | ArmCL 18.03 (measured)
-|-|-|-
MobileNet_v1_1.0_224 | 89.500 | 89.410 | 89.398 |
MobileNet_v1_1.0_192 | 88.900 | 88.466 | 88.474 |
MobileNet_v1_1.0_160 | 87.500 | 87.084 | 87.088 |
MobileNet_v1_1.0_128 | 85.300 | 84.928 | 84.940 |

**Comment by psyhtest** [comment]
We first noticed accuracy degradation (up to 0.5%) when comparing the ArmCL results with the TensorFlow accuracy figures _claimed_ by Google researchers. However, when we _measured_ the TensorFlow accuracy ourselves using the same methodology as for ArmCL, it became clear that degradation is probably due to differences in input preprocessing (e.g. resizing to a larger input resolution and then cropping to the model resolution; cf. with the [ReQuEST artifact from Intel](https://github.com/ctuning/ck-request-asplos18-caffe-intel/issues/7#issuecomment-367319271)).

**Comment by psyhtest** [comment]
Now, a fraction of a percent may not sound like much (e.g. 0.002% means just 1 image out of 50,000, 0.006% - 3 images, etc.) but this is smoothed out over 50,000 images. In fact, for the first 500 images, our ArmCL implementation disagrees with TensorFlow about 3 images:

1. `ILSVRC2012_val_00000060.JPEG`
```
$ ck run program:classification-tensorflow \
--env.CK_IMAGE_FILE=/home/anton/CK_TOOLS/dataset-imagenet-ilsvrc2012-val-min/ILSVRC2012_val_00000060.JPEG

---------------------------------------
ILSVRC2012_val_00000060.JPEG - (588) n03482405 hamper
0.47 - (588) n03482405 hamper
0.43 - (492) n03014705 chest
0.02 - (626) n03666591 lighter, light, igniter, ignitor
0.01 - (526) n03179701 desk
0.01 - (681) n03832673 notebook, notebook computer
---------------------------------------

$ ck run program:mobilenets-armcl-opencl \
--env.CK_IMAGE_FILE=/home/anton/CK_TOOLS/dataset-imagenet-ilsvrc2012-val-min/ILSVRC2012_val_00000060.JPEG

---------------------------------------
ILSVRC2012_val_00000060.JPEG - (588) n03482405 hamper
0.45 - (492) n03014705 chest
0.45 - (588) n03482405 hamper
0.02 - (626) n03666591 lighter, light, igniter, ignitor
0.01 - (526) n03179701 desk
0.01 - (681) n03832673 notebook, notebook computer
---------------------------------------

```

**Comment by AnthonyARM** [comment]
Hi Anton,
I only went quickly through the "MobileNets using ArmCL" repo as I was triaging this bug and didn't find the answer: how is ACL integrated with TensorFlow to run the graphs ? Is it using the graph API ? Hand written networks ? Through Android NN ?

**Comment by psyhtest** [comment]
hi @AnthonyARM, the ArmCL based implementation is built on top of your [graph API example](https://github.com/dividiti/ck-request-asplos18-mobilenets-armcl-opencl/blob/master/program/mobilenets-armcl-opencl/mobilenets.cpp). The TensorFlow implementation is a separate [Python program](https://github.com/ctuning/ck-tensorflow/blob/master/program/classification-tensorflow/classify.py) (which uses exactly the same pretrained weights though).

**Comment by gmiodice** [comment]
Hi @psyhtest 

many thanks for reporting this. Do you know if you have the same discrepancy for both NEON and CL backend?

Thanks

**Comment by psyhtest** [comment]
Please find ArmCL instructions below. (TensorFlow instructions to follow shortly.)

Please me know if anything is unclear (or ask a friendly person from @sztaylor's team:)).

## Installing artifact dependencies
```
$ sudo apt install python python-pip
$ sudo apt install libblas-dev liblapack-dev libatlas-base-dev
$ sudo python-numpy python-scipy
$ sudo pip install pillow
$ sudo pip install ck
```

## Detecting GCC, Python, OpenCL
```
$ ck detect soft:compiler.gcc
$ ck detect soft:compiler.python
$ ck detect platform.gpgpu --opencl
```

## Installing the MobileNets artifact
```
$ ck pull repo --url=https://github.com/dividiti/ck-request-asplos18-mobilenets-armcl-opencl
$ ck install ck-caffe:package:imagenet-2012-aux
$ ck install ck-caffe:package:imagenet-2012-val-min-resized
$ ck install ck-math:package:lib-armcl-opencl-18.03 --env.USE_GRAPH=ON --env.USE_NEON=ON
$ ck install ck-request-asplos18-mobilenets-armcl-opencl:package:weights-mobilenet-v1-1.0-224-npy
```

## Running the classification program
To classify `ILSVRC2012_val_00000001.JPEG`:
```
$ ck compile program:mobilenets-armcl-opencl
$ ck run program:mobilenets-armcl-opencl
```

To classify another image (e.g. `ILSVRC2012_val_00000060.JPEG`):
```
$ ck virtual env --tags=imagenet,small_dataset
$ ck run program:mobilenets-armcl-opencl \
--env.CK_IMAGE_FILE=$CK_ENV_DATASET_IMAGENET_VAL/ILSVRC2012_val_00000060.JPEG
...
  (run ...)
executing code ...
Kernel path: /home/anton/CK_TOOLS/lib-armcl-opencl-18.03-gcc-6.3.0-linux-64/src/src/core/CL/cl_kernels/
Image list file: ../images-224-1-1.txt
Image count in file: 1
Batch list file: ../batches-224-1-1.txt
Batch count in file: 1

Prepare graph...

Run graph...

Batch 1 of 1
File: ../batches-224-1-1/ILSVRC2012_val_00000060.JPEG.npy
Loaded in 0.0215138 s
Classified in 0.0713635s
Test passed
-------------------------------
Graph loaded in 1.36456 s
All batches loaded in 0.0215138 s
All batches classified in 0.0713635 s
Average classification time: 0.0713635 s
-------------------------------

  (post processing: "python /home/anton/CK_REPOS/ck-request-asplos18-mobilenets-armcl-opencl/program/mobilenets-armcl-opencl/postprocess.py"

--------------------------------
Process results in predictions
---------------------------------------
ILSVRC2012_val_00000060.JPEG - (588) n03482405 hamper
0.45 - (492) n03014705 chest
0.45 - (588) n03482405 hamper
0.02 - (626) n03666591 lighter, light, igniter, ignitor
0.01 - (526) n03179701 desk
0.01 - (681) n03832673 notebook, notebook computer
---------------------------------------
Accuracy top 1: 0.000000 (0 of 1)
Accuracy top 5: 1.000000 (1 of 1)
--------------------------------

  (reading fine grain timers from tmp-ck-timer.json ...)

{
  "accuracy_top1": 0.0,
  "accuracy_top5": 1.0,
  "execution_time": 0.071363,
  "execution_time_sum": 1.45744,
  "frame_predictions": [
    {
      "accuracy_top1": "no",
      "accuracy_top5": "yes",
      "class_correct": 588,
      "class_topmost": 492,
      "file_name": "ILSVRC2012_val_00000060.JPEG"
    }
  ],
  "images_load_time_avg_s": 0.021514,
  "images_load_time_s": 0.021514,
  "prediction_time_avg_s": 0.071363,
  "prediction_time_total_s": 0.071363,
  "setup_time_s": 1.364563,
  "test_time_s ": 0.103095
}

Execution time: 0.071 sec.
```

**Comment by psyhtest** [comment]
@gmiodice I am afraid I don't know. Our ArmCL program currently supports only OpenCL. It should not be hard to extend it to support Neon.

**Comment by psyhtest** [comment]
Please find TensorFlow instructions below (assuming CK has been installed as per ArmCL instructions [above](https://github.com/ARM-software/ComputeLibrary/issues/431#issuecomment-381920505).)

## Installing artifact dependencies
```
$ sudo apt install liblapack-dev libatlas-dev
$ sudo pip install enum34 mock pillow wheel absl-py scipy
$ ck install ck-env:package:tool-bazel-0.11.1-linux
```

## Installing the artifact
```
$ ck pull repo:ck-tensorflow
$ ck install package:lib-tensorflow-1.7.0-src-cpu
$ ck install package:tensorflowmodel-mobilenet-v1-1.0-224-py
```

## Running the classification program

To classify e.g. `ILSVRC2012_val_00000060.JPEG`:
```
$ ck virtual env --tags=imagenet,small_dataset
$ ck run program:classification-tensorflow \
--env.CK_IMAGE_FILE=$CK_ENV_DATASET_IMAGENET_VAL/ILSVRC2012_val_00000060.JPEG
```
